### PR TITLE
enable multicluster sync for groups

### PR DESCRIPTION
**Pull request by RolandMa1986**
Signed-off-by: Roland.Ma <rolandma@yunify.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Groups are needed to be synced to member clusters when the MultiCluster feature is enabled. So member clusters can be authorized with group permission.

**Which issue(s) this PR fixes**:

### CODE REVIEW

File: pkg/controller/group/group_controller.go [hunk 1]
```diff
@@ -192,3 +236,57 @@ func (c *Controller) deleteRoleBindings(group *iam1alpha2.Group) error {
 
 	return nil
 }
+
+func (c *Controller) multiClusterSync(group *iam1alpha2.Group) error {
+
+	obj, err := c.federatedGroupLister.Get(group.Name)
+	if obj == nil {
```
Comment (wansir) [inline pkg/controller/group/group_controller.go:243]: Internal server error should be returned first, create the federation object if the error is `not found`.

File: pkg/controller/group/group_controller.go [hunk 2]
```diff
@@ -192,3 +236,57 @@ func (c *Controller) deleteRoleBindings(group *iam1alpha2.Group) error {
...
+	}
+	if err != nil {
+		klog.Error(err)
+		return err
+	}
+
+	if !reflect.DeepEqual(obj.Spec.Template.Labels, group.Labels) ||
+		!reflect.DeepEqual(obj.Spec.Template.Annotations, group.Annotations) {
+
+		obj.Spec.Template.Annotations = group.Annotations
```
Comment (wansir) [inline pkg/controller/group/group_controller.go:254]: The annotations field is [managed by controllers in member clusters](https://github.com/kubernetes-sigs/kubefed/blob/master/docs/userguide.md#local-value-retention). If the field must be synchronized in multiple clusters, we should use [overrides](https://github.com/kubernetes-sigs/kubefed/blob/master/docs/userguide.md#overrides)

File: pkg/controller/group/group_controller.go (hunk 1 above)
Comment (RolandMa1986) [reply pkg/controller/group/group_controller.go]: Refined.

File: pkg/controller/group/group_controller.go (hunk 2 above)
Comment (RolandMa1986) [reply pkg/controller/group/group_controller.go]: The annotations field is not necessary, removed it.

File: pkg/controller/group/group_controller.go [hunk 3]
```diff
@@ -106,10 +125,27 @@ func (c *Controller) reconcile(key string) error {
...
+		if c.multiClusterEnabled {
+			// Ensure not controlled by Kubefed
+			if group.Labels == nil || group.Labels[constants.KubefedManagedLabel] != "false" {
+				if g == nil {
+					g = group.DeepCopy()
+				}
+				if g.Labels == nil {
+					g.Labels = make(map[string]string, 0)
+				}
+				g.Labels[constants.KubefedManagedLabel] = "false"
```
Comment (zryfish) [inline pkg/controller/group/group_controller.go:143]: So what if the group doesn't have a `KubefedManagedLabel`, we are going to add a label to it.  I think we'd better do nothing If the group has no label `KubefedManagedLabel`. The L135 
```
    if group.Labels != nil && group.Labels[constants.KubefedManagedLabel] == "true" {
        group.Labels[constants.KubefedManagedLabel] = "false"
    }
```

File: pkg/controller/groupbinding/groupbinding_controller.go [hunk 4]
```diff
@@ -101,14 +119,32 @@ func (c *Controller) reconcile(key string) error {
...
 		if !sliceutil.HasString(groupBinding.Finalizers, finalizer) {
-			groupBinding.ObjectMeta.Finalizers = append(groupBinding.ObjectMeta.Finalizers, finalizer)
-			if groupBinding, err = c.ksClient.IamV1alpha2().GroupBindings().Update(groupBinding); err != nil {
+			g = groupBinding.DeepCopy()
+			g.ObjectMeta.Finalizers = append(g.ObjectMeta.Finalizers, finalizer)
+		}
+
+		if c.multiClusterEnabled {
+			// Ensure not controlled by Kubefed
+			if groupBinding.Labels == nil || groupBinding.Labels[constants.KubefedManagedLabel] != "false" {
```
Comment (zryfish) [inline pkg/controller/groupbinding/groupbinding_controller.go:130]: same here

File: pkg/controller/group/group_controller.go (hunk 3 above)
Comment (RolandMa1986) [reply pkg/controller/group/group_controller.go:143]: Group will be labeled with KubefedManagedLabel by Kubefed when multi-cluster enabled.  We must prevent Kubefed change Group on the Host cluster by the label.

Comment (zryfish) [reply pkg/controller/group/group_controller.go:143]: yes, but what if the group without KubefedManagedLabel, why do we still add a KubefedManagedLabel=false to it?

Comment (RolandMa1986) [reply pkg/controller/group/group_controller.go:143]: This Label will be added only when multi-cluster enabled and on the Master cluster, so it must exist. Immediately or to be updated by Kubefed when FedeatedGroup created.

**Pull request merged**